More aggressive testing of inconsistent record reads, and fixing of surfaced issues

chrisvest opened the pull request:
We are not allowed to throw exceptions due to record decoding errors until we have verified that our read access to a page was consistent, wrt. page locks and eviction.

This is solved by introducing a _cursor exception_ feature of the page cursors, whereby the decoding code can attach exceptions to the cursor during the decoding of records. We will then check for the presence of such an exception after it has been verified that the page access was consistent.

This PR also introduce a much more aggressive testing of inconsistent reads, by making the `AdversarialReadPageCursor` more devious: Instead of returning zeros to the reading code upon an injected inconsistent read, the adversarial cursor will instead count the operations performed upon it, and choose one at random. When that operation is later retried, the adversarial cursor will return random data from that point on. This allows more spectacular bugs to show up. These bugs are also fixed in this PR.

CHANGELOG:
Fix rare record decoding exceptions caused by page reads racing with page writes to the same record, or page eviction.

chrisvest committed:
More aggressive inconsistent reads from the PageCacheRule

We previously just returned zeros from the adversarial page read cursors when doing an inconsistent read.
Now, we count the number of operations performed against the page cursor, choose one, tell the reader to retry, and then we return random garbage data from the point of the chosed operation.
This finds many, many more cases of faulty use of the page cursors.

chrisvest committed:
Fix rare inconsistent read problems surfaced by the more devious PageCacheRule

chrisvest committed:
Fix DatabaseRebuildToolTest that broke because we changed the implementation of `PropertyRecord.toString()`

chrisvest committed:
Update PageCursor javadocs

chrisvest committed:
Add a PageCursor.setCursorError and .checkAndClearCursrorError feature

This will help simplify the design around how consistent record reads report errors up the stack.

chrisvest committed:
Make the record formats report decoding errors through the page cursor setCursorError mechanism.

chrisvest commented:
@MishaDemianenko Take a look at these recent changes.

MishaDemianenko commented on community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialReadPageCursor.java:
```diff
@@ -286,14 +413,92 @@ public boolean checkAndClearBoundsFlag()
+            this.offset = offset;
+            this.insteadOf = insteadOf;
+            trace = new Exception( toString() );
+            trace.fillInStackTrace();
```
why do we call fillInStackTrace here ourselves? constructor just did it for us

MishaDemianenko commented on community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialReadPageCursor.java:
```diff
@@ -286,14 +413,92 @@ public boolean checkAndClearBoundsFlag()
+        for ( Object o : s.inconsistentReadHistory )
+        {
+            sb.append( o.toString() ).append( '\n' );
+            if ( o instanceof NumberValue )
```
why do we need to do cast here at all and invoke type specific `printStackTrace` method instead of just do it as part of NumberValue toString call?

chrisvest replied:
The toString is for making things readable in the debugger. Adding the stack trace would add too much clutter.

chrisvest committed:
Small cleanup in AdversarialReadPageCursor

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/standard/PropertyRecordFormat.java:
```diff
@@ -78,7 +78,18 @@ public void read( PropertyRecord record, PageCursor cursor, RecordLoad mode, int
+                return;
+            }
+            int additionalBlocks = numberOfBlocksUsed - 1;
+            if ( additionalBlocks * 8 > RECORD_SIZE - (cursor.getOffset() - offsetAtBeginning ) )
```
looks like a place for Byte.SIZE ?

MishaDemianenko replied:
or even better do << 3 instead ?

MishaDemianenko commented on community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialReadPageCursor.java:
```diff
@@ -232,16 +342,18 @@ public void rewind()
     public boolean next() throws IOException
     {
-        currentReadIsInconsistent = adversary.injectFailureOrMischief( FileNotFoundException.class, IOException.class,
+        boolean currentReadIsPreparingInconsistent = state.injectFailureOrMischief( FileNotFoundException.class, IOException.class,
```
state.injectFailureOrMischief and state.reset always come together maybe it makes sense to introduce method in state that will do both things? without external state touch

MishaDemianenko commented on community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialReadPageCursor.java:
```diff
@@ -255,24 +367,39 @@ public void close()
-        adversary.injectFailure( IndexOutOfBoundsException.class );
-        if ( !currentReadIsInconsistent )
+        state.injectFailure( IndexOutOfBoundsException.class );
+        if ( state.currentReadIsPreparingInconsistent )
```
this if looks like state abstraction gives a leak and require external manipulation over internal state

MishaDemianenko commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/MuninnPageCursor.java:
```diff
@@ -158,6 +172,8 @@ void clearPageState()
         pageSize = 0; // make all future bound checks fail
         page = null; // make all future page navigation fail
+        currentPageId = UNBOUND_PAGE_ID;
+        cursorException = null;
```
wdyt about just checking cursorException is really null and throw exception otherwise instead? to completely avoid cases of lost exception

chrisvest replied:
I can add an assert about that.

chrisvest replied:
On a second thought, I don't really want to do that. I don't feel that it's part of the page cursors responsibility. The cursor exceptions are managed by the code that uses the cursor, so I don't want to let the cursor assume that it knows what the exceptions mean.

MishaDemianenko commented on community/io/src/main/java/org/neo4j/io/pagecache/PageCursor.java:
```diff
@@ -297,23 +293,49 @@
+     *
+     * @param message The message of the {@link CursorException} that {@link #checkAndClearCursorError()} will throw.
+     */
+    public abstract void setCursorError( String message );
```
wdyt about renaming `setCursorError` and `clearCursorError` to `setCursorException` and `clearCursorException` while i'm reading `error` i'm really expection Error to be used for that then instead of exception and it's a bit confusing.

chrisvest replied:
Okay.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/record/RecordLoad.java:
```diff
@@ -39,77 +40,66 @@
-            return true;
+            try
+            {
+                cursor.checkAndClearCursorError();
```
same here with Error -> Exception thingy

MishaDemianenko commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/MuninnPageCursor.java:
```diff
@@ -37,6 +39,9 @@
     private static final boolean tracePinnedCachePageId =
             flag( MuninnPageCursor.class, "tracePinnedCachePageId", false );
 
+    private static final boolean usePreciseCursorErrorStackTraces =
```
do you think we will ever need non precise stack traces? what will be use case of that?
because of this complication we need to declare `cursorException` to be Object, and significantly complicate all code around that

chrisvest replied:
I was thinking of turning them off by default, or maybe have it off in EE. It might produce a slight bit of overhead, but I don't know if its noticeable.

chrisvest replied:
How about we turn it off by default, and enable it in testing?

MishaDemianenko commented:
@chrisvest i have finished my first round of review, couple of comments

chrisvest committed:
Rename cursor error to cursor exception

- checkAndClearCursorError -> checkAndClearCursorException
- setCursorError -> setCursorException
- clearCursorError -> clearCursorException

chrisvest committed:
Cleanups in AdversarialReadPageCursor

chrisvest committed:
Disable precise cursor exception stack traces by default, but enable it in testing
